src/metaswitch/clearwater/queue_manager/queue_fsm.py

```diff
@@ -0,0 +1,138 @@
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import constants
+from .queue_state import QueueStateInfo
+from .alarms import QueueAlarm
+from .timers import QueueTimer
+import logging
+from functools import partial
+from queue_utils import move_to_processing, mark_node_as_errored
+
+_log = logging.getLogger("queue_manager.queue_fsm")
+
+class QueueFSM(object):
+
+    def __init__(self, plugin, node_id, callback_func):
+        self._queue_states = None
+        self._queue_config = {}
+        self._last_local_state = None
+
+        self._plugin = plugin
+        self._id = node_id
+        self._running = True
+
+        self._timer_callback_func = callback_func
+        self._timer = None
+        self._local_alarm = QueueAlarm(*self._plugin.local_alarm())
+        self._global_alarm = QueueAlarm(*self._plugin.global_alarm())
+
+        self._local_fsm = {constants.LS_NO_QUEUE: [self._local_alarm.clear],
+                           constants.LS_NO_QUEUE_ERROR: [self._local_alarm.critical],
+                           constants.LS_FIRST_IN_QUEUE: [self._local_alarm.minor,
+                                                         partial(move_to_processing, self.return_queue_config, self._id)],
+                           constants.LS_PROCESSING: [self._plugin.at_front_of_queue,
+                                                     self._set_timer_with_id],
+                           constants.LS_WAITING_ON_OTHER_NODE: [self._local_alarm.clear,
+                                                                self._set_timer_with_current_node_id],
+                           constants.LS_WAITING_ON_OTHER_NODE_ERROR: [self._local_alarm.critical,
+                                                                      self._set_timer_with_current_node_id]}
+
+        self._global_fsm = {constants.GS_NO_SYNC: self._global_alarm.clear,
```

Comment:
Actually, thinking about it harder - I don't think the global state actually changes the state transitions (c.f. the cluster manager, where the local and global states are both inputs to the state machine, and the same local state might trigger two different actions based on the global state). Here, for a given local state, I think you'll always carry out the same action, and the global state is just something we trigger alarms off - so it might be clearer to represent it as something other than an FSM.